File: utilities/calibration/04_daily_activity_pattern_TM.R
library(dplyr)
library(tidyr)
options(java.parameters = "-Xmx8000m")  # xlsx uses java and can run out of memory
library(xlsx)

# For RStudio, these can be set in the .Rprofile
TARGET_DIR   <- Sys.getenv("TARGET_DIR")  # The location of the input files
ITER         <- Sys.getenv("ITER")        # The iteration of model outputs to read
SAMPLESHARE  <- Sys.getenv("SAMPLESHARE") # Sampling
CODE_DIR     <- Sys.getenv("CODE_DIR")    # location of utilitiles\calibration code
CALIB_ITER   <- Sys.getenv("CALIB_ITER")  # calibration iteration

TARGET_DIR   <- gsub("\\\\","/",TARGET_DIR) # switch slashes around
CODE_DIR     <- gsub("\\\\","/",CODE_DIR  ) # switch slashes around
OUTPUT_DIR   <- file.path(TARGET_DIR, paste0("OUTPUT_",CALIB_ITER), "calibration")
if (!file.exists(OUTPUT_DIR)) { dir.create(OUTPUT_DIR) }

stopifnot(nchar(TARGET_DIR  )>0)
stopifnot(nchar(ITER        )>0)
stopifnot(nchar(SAMPLESHARE )>0)
stopifnot(nchar(CODE_DIR    )>0)
stopifnot(nchar(CALIB_ITER  )>0)

SAMPLESHARE <- as.numeric(SAMPLESHARE)

print(paste0("TARGET_DIR  = ",TARGET_DIR ))
print(paste0("ITER        = ",ITER       ))
print(paste0("SAMPLESHARE = ",SAMPLESHARE))
print(paste0("CODE_DIR    = ",CODE_DIR   ))
print(paste0("CALIB_ITER  = ",CALIB_ITER ))

WORKBOOK       <- "M:\\Development\\Travel Model One\\Calibration\\Version 1.5.2\\04 Coordinated Daily Activity Pattern\\04_CoordinatedDailyActivityPattern.xlsx"
WORKBOOK       <- gsub("\\\\","/",WORKBOOK  ) # switch slashes around
WORKBOOK_TEMP  <- gsub(".xlsx","_temp.xlsx", WORKBOOK)
WORKBOOK_BLANK <- file.path(CODE_DIR, "workbook_templates", "04_CoordinatedDailyActivityPattern_blank.xlsx")
calib_workbook <- loadWorkbook(file=WORKBOOK_BLANK)
calib_sheets   <- getSheets(calib_workbook)

cdap_results <- read.table(file=file.path(TARGET_DIR,paste0("OUTPUT_",CALIB_ITER),"main","cdapResults.csv"), header=TRUE, sep=",")

# summarize to (county, Auto Ownership)
cdap_ptype  <- group_by(cdap_results, PersonType, ActivityString) %>% summarise(num_pers=n())
# divide by SAMPLESHARE
cdap_ptype  <- mutate(cdap_ptype, num_pers=num_pers/SAMPLESHARE)

cdap_ptype_spread <- spread(cdap_ptype, key=ActivityString, value=num_pers)

# save it
outfile <- file.path(OUTPUT_DIR, paste0("04_daily_activity_pattern_TM.csv"))
write.table(cdap_ptype_spread, outfile, sep=",", row.names=FALSE)
print(paste("Wrote",outfile))

# save it (along with source label) to calibration workbook
addDataFrame(as.data.frame(cdap_ptype_spread), calib_sheets$modeldata, startRow=2, startColumn=1, row.names=FALSE)
source_cell <- getCells( getRows(calib_sheets$modeldata, rowIndex=1:1), colIndex=1:1 )
setCellValue(source_cell[[1]], paste("Source: ",outfile))

saveWorkbook(calib_workbook, WORKBOOK_TEMP)
forceFormulaRefresh(WORKBOOK_TEMP, WORKBOOK, verbose=TRUE)
print(paste("Wrote",WORKBOOK))

File: utilities/calibration/04_daily_activity_pattern_TM.py
import argparse
import pandas as pd
import numpy as np
import os
import sys
from pathlib import Path

# Import the calibration framework  
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from calibration_framework import CalibrationBase, SheetTarget, add_county_info
from calibration_data_models import (
    CTRAMPPersonType,
    CDAPSummary,
    CDAPSummaryBATS,
    validate_dataframe
)


class DailyActivityPatternCalibration(CalibrationBase):
    """Calibration processor for coordinated daily activity pattern."""
    
    # sheet, column, startRow, endRow
    # used for populating current iteration constants in calibration workbook with model UEC input
    # UEC workbook is ".xls" and uses xlrd which is 0-based indexing for rows/columns, but
    # config uses 1-based for readability, so offsets are applied in the reading functions
    UEC_SOURCE_RANGES = {
        "Mandatory":        ("OnePerson",  7, 33, 40),
        "Non-Mandatory":    ("OnePerson", 8, 33, 40),
        "Home":             ("OnePerson", 9, 33, 40),
    }
    # calibration workbook is ".xlsx" and uses openpyxl which is 1-based indexing for rows/columns
    CALIBRATION_DESTINATION_RANGES = {
        "Mandatory":        ("calibration", 3, 31, 38),
        "Non-Mandatory":    ("calibration", 4, 31, 38),
        "Home":             ("calibration", 5, 31, 38),

    }
    

    def __init__(self, config_file: str = None):
        super().__init__("04", config_file)
        self.bats_data = self.submodel_config.get("bats_data", False)
    
    def process_data(self) -> dict:
        """Process the coordinated daily activity pattern data."""
        # Load input data
        sep = "=" * 80
        self.logger.info(f"\n{sep}\nPROCESS DATA\n{sep}")
        cdap_results = pd.read_csv(self.submodel_config['input_file'])
        self.logger.info(f"Reading in CDAP Input File: {self.submodel_config['input_file']}")
        CTRAMPPersonTypeLookUp = {person.label: person.value for person in CTRAMPPersonType}


        if self.bats_data :
            cdap_results['person_type'] = cdap_results['type'].map(CTRAMPPersonTypeLookUp)
            cdap_results.fillna({'person_weight':0}, inplace = True)

            # Home TAZ comes from household data; attach home county for filtering.
            taz_data = pd.read_csv(self.config.get('data_sources', 'taz_data'), usecols=['ZONE', 'COUNTY'])
            hh = pd.read_csv(self.config.get('data_sources', 'household_file'), usecols=['hh_id', 'taz'])
            cdap_results = cdap_results.merge(hh, on='hh_id', how='left')
            cdap_results = add_county_info(cdap_results, taz_data, self.county_lookup,
                                           taz_col='taz', county_col_name='home_county_id', county_name_col='home_county')

            # One summary per county filter ("all" = unfiltered); filter on home county only.
            results = {}
            for label, subset in self.iter_county_filters(cdap_results, orig_county_col="home_county", dest_county_col=None):
                cdap_ptype = subset.groupby(['person_type', 'activity_pattern'])['person_weight'].sum().reset_index(name='num_pers')
                results[f'person_type_summary_{label}'] = cdap_ptype
            return results

        else:
        # Summarize by person type and activity string
            cdap_ptype = cdap_results.groupby(['PersonType', 'ActivityString']).size().reset_index(name='num_pers')
            cdap_ptype['num_pers'] = cdap_ptype['num_pers'] / self.sampleshare
            
            # Pivot to spread format
            cdap_ptype_spread = cdap_ptype.pivot(index='PersonType', columns='ActivityString', values='num_pers')
            cdap_ptype_spread = cdap_ptype_spread.fillna(0).reset_index()
            
            return {
                'person_type_summary': cdap_ptype_spread
            }
    
    def validate_outputs(self, results:dict):
        """Validate outputs before generating the files and updating excel"""
        sep = "=" * 80
        self.logger.info(f"\n{sep}\nOUTPUT VAlIDATION\n{sep}")

        if self.bats_data:
            for key, df in results.items():
                if key.startswith('person_type_summary_') and df is not None:
                    validate_dataframe(df, CDAPSummaryBATS)
            self.logger.info("Person Type BATS Summary Validated (all county filters)")
        elif results['person_type_summary'] is not None:
            validate_dataframe(results['person_type_summary'], CDAPSummary)
            self.logger.info("Person Type Summary Validated")


    def generate_outputs(self, results: dict):
        """Generate output files and Excel updates."""

        sep = "=" * 80
        self.logger.info(f"\n{sep}\nGENERATE OUTPUTS\n{sep}")

        if self.bats_data:
            targets = [SheetTarget("person_type_summary_all", "BATS 2023", 2, 2,
                                   "dap_summaries.csv", (1, 2))]
            for key in results:
                if key.endswith("_all"):
                    continue
                targets.append(SheetTarget(key, "", 0, 0, f"BATS2023_{key}.csv", None))
        else:
            targets = [SheetTarget("person_type_summary", "modeldata", 2, 1,
                                   f"{self.submodel}_daily_activity_pattern_TM.csv", (1, 1))]
        self.write_results_to_workbook(results, targets)



def main():
    """Main entry point for the daily activity pattern calibration."""
    parser = argparse.ArgumentParser(description="Coordinated daily activity pattern calibration")
    parser.add_argument("--config", default=None, help="Path to calibration_config.yaml (default: same directory as this script)")
    args = parser.parse_args()

    calibration = DailyActivityPatternCalibration(config_file=args.config)
    calibration.run()


if __name__ == "__main__":
    main()
